EssentialApp/EssentialAppTests/Image Comments/ImageCommentsUIIntegrationTests.swift

```diff
@@ -0,0 +1,227 @@
+		
+		sut.refreshControl?.simulatePullToRefresh()
+		XCTAssertEqual(sut.errorMessage, nil)
+	}
+	
+	func test_onBackButtonPressed_cancelRequestFeedFromLoader() {
+		let loader = LoaderSpy()
+		var sut: ImageCommentsViewController?
+		
+		autoreleasepool {
+			sut = ImageCommentsUIComposer.imageCommentsComposedWith(loader: loader.loadPublisher)
+			sut?.loadViewIfNeeded()
+		}
+
+		XCTAssertEqual(loader.cancelCount, 0)
+
+		sut = nil
+		XCTAssertEqual(loader.cancelCount, 1)
+	}
+	
+	func test_loadImageCommentsCompletion_dispatchesFromBackgroundToMainThread() {
+		let (sut, loader) = makeSUT()
+		sut.loadViewIfNeeded()
+		
+		let exp = expectation(description: "Wait for background queue")
+		DispatchQueue.global().async {
+			loader.completeLoading(at: 0)
+			exp.fulfill()
+		}
+		wait(for: [exp], timeout: 1.0)
+	}
+	
+	//MARK: Helpers
+	
+	private func makeSUT(file: StaticString = #file, line: UInt = #line) -> (sut: ImageCommentsViewController, loader: LoaderSpy) {
+		let loader = LoaderSpy()
+		let sut = ImageCommentsUIComposer.imageCommentsComposedWith(loader: loader.loadPublisher)
+		
+		trackForMemoryLeaks(sut, file: file, line: line)
+		trackForMemoryLeaks(loader, file: file, line: line)
+		
+		return (sut, loader)
+	}
+	
+	private func makeComment(id: UUID, message: String, created_at: Date, username: String) -> ImageComment {
+		let author = CommentAuthor(username: username)
+		let comment = ImageComment(id: id, message: message, createdDate: created_at, author: author)
+		
+		return comment
+	}
+	
+	private func relativeDateStringFromNow(to date: Date) -> String {
+		let formatter = RelativeDateTimeFormatter()
+		formatter.unitsStyle = .full
+		let relativeDateString = formatter.localizedString(for: date, relativeTo: Date())
+		return relativeDateString
+	}
+	
+	private func assertThat(_ sut: ImageCommentsViewController, isRendering imageComments: [ImageComment], file: StaticString = #file, line: UInt = #line) {
+		XCTAssertEqual(sut.numberOfRenderedImageCommentsViews, imageComments.count, file: file, line: line)
+		
+		imageComments.enumerated().forEach { index, comment in
+			let cell = sut.renderedCell(at: index)
+			XCTAssertEqual(cell?.usernameText, comment.author.username, file: file, line: line)
+			XCTAssertEqual(cell?.createdTimetext, relativeDateStringFromNow(to: comment.createdDate), file: file, line: line)
+			XCTAssertEqual(cell?.messageText, comment.message, file: file, line: line)
+		}
+	}
+	
+	func localized(_ key: String, file: StaticString = #filePath, line: UInt = #line) -> String {
+		let table = "ImageComments"
+		let bundle = Bundle(for: ImageCommentsPresenter.self)
+		let value = bundle.localizedString(forKey: key, value: nil, table: table)
+		if value == key {
+			XCTFail("Missing localized string for key: \(key) in table: \(table)", file: file, line: line)
+		}
+		return value
+	}
+	
+	class LoaderSpy: ImageCommentsLoader {
+		var cancelCount = 0
```

Comment:
The setter for this property should be private because clients shouldn't be able to change it.

```suggestion
		private(set) var cancelCount = 0
```